### PR TITLE
Iterant fold left operators need to handle errors thrown in F[_] context

### LINKED ISSUE #563: Iterant fold left operators (yielding F[_]) need to handle errors thrown in F[_] context
See #559

### PR DESCRIPTION
**Pull request by Avasil**
Fixes #563

### CODE REVIEW

**Comment by Avasil** [comment]
I did benchmark for```Iterant#toListL``` to satisfy my curiosity:
```scala
Before:
[info] Benchmark                                 Mode  Cnt       Score      Error  Units
[info] IterantFoldLeftLBenchmark.toListL100     thrpt  100  391092.509 ± 3795.786  ops/s
[info] IterantFoldLeftLBenchmark.toListL10000   thrpt  100    4020.913 ±   30.503  ops/s
[info] IterantFoldLeftLBenchmark.toListL100000  thrpt  100     425.652 ±    1.370  ops/s

After:
[info] Benchmark                                 Mode  Cnt       Score      Error  Units
[info] IterantFoldLeftLBenchmark.toListL100     thrpt  100  361662.963 ± 4809.333  ops/s
[info] IterantFoldLeftLBenchmark.toListL10000   thrpt  100    3952.470 ±   48.976  ops/s
[info] IterantFoldLeftLBenchmark.toListL100000  thrpt  100     376.399 ±    3.289  ops/s
```

File: monix-tail/shared/src/main/scala/monix/tail/internal/IterantFoldLeft.scala [hunk 1]
```diff
@@ -31,37 +33,51 @@ private[tail] object IterantFoldLeft {
   final def apply[F[_], S, A](source: Iterant[F, A], seed: => S)(op: (S,A) => S)
     (implicit F: Sync[F]): F[S] = {
 
-    def loop(self: Iterant[F, A], state: S): F[S] = {
+    def loop(stopRef: ObjectRef[F[Unit]])(self: Iterant[F, A], state: S): F[S] = {
```
Comment (alexandru) [inline monix-tail/shared/src/main/scala/monix/tail/internal/IterantFoldLeft.scala]: Lets do currying here. Put the `state` on the left, such that you can then do ... 

```scala
rest.flatMap(loop(stopRef, newState))
```

File: monix-tail/shared/src/main/scala/monix/tail/internal/IterantFoldWhileLeft.scala [hunk 2]
```diff
@@ -45,26 +47,30 @@ private[tail] object IterantFoldWhileLeft {
       if (hasResult)
         stop.map(_ => s)
       else
-        rest.flatMap(loop(s))
+        rest.flatMap(loop(stopRef)(s))
     }
 
-    def loop(state: S)(self: Iterant[F, A]): F[S] = {
+    def loop(stopRef: ObjectRef[F[Unit]])(state: S)(self: Iterant[F, A]): F[S] = {
```
Comment (alexandru) [inline monix-tail/shared/src/main/scala/monix/tail/internal/IterantFoldWhileLeft.scala]: We don't need 3 parameter groups for this curried function. `stopRef` and `state` can be on the same level so to speak.

File: monix-tail/shared/src/main/scala/monix/tail/internal/IterantFoldWhileLeft.scala [hunk 3]
```diff
@@ -45,26 +47,30 @@ private[tail] object IterantFoldWhileLeft {
...
+          process(stopRef)(state, cursor, rest, stop)
 
-        case Suspend(rest, _) =>
-          rest.flatMap(loop(state))
+        case Suspend(rest, stop) =>
+          stopRef.elem = stop
+          rest.flatMap(loop(stopRef)(state))
 
         case Last(a) =>
           F.pure(f(state, a) match {
```
Comment (alexandru) [inline monix-tail/shared/src/main/scala/monix/tail/internal/IterantFoldWhileLeft.scala]: Calling `f` can throw, which can trigger an exception that's going to trigger our `handleErrorWith`, which is going to call an `earlyStop` reference that's no longer in service.

Do `stopRef.elem = null`.

File: monix-tail/shared/src/main/scala/monix/tail/internal/IterantFoldWhileLeft.scala [hunk 4]
```diff
@@ -74,55 +80,69 @@ private[tail] object IterantFoldWhileLeft {
...
           }
 
-        case Suspend(rest, _) =>
-          rest.flatMap(loop(state))
+        case Suspend(rest, stop) =>
+          stopRef.elem = stop
+          rest.flatMap(loop(stopRef)(state))
 
         case Last(a) =>
           f(state, a).map {
```
Comment (alexandru) [inline monix-tail/shared/src/main/scala/monix/tail/internal/IterantFoldWhileLeft.scala]: Same problem here. `stopRef.elem = null`. If unsure then it's easier to think that this assignment needs to happen on all branches, unless you can reason about why it isn't needed.

File: monix-tail/shared/src/main/scala/monix/tail/internal/IterantFoldWhileLeft.scala (hunk 3 above)
Comment (Avasil) [reply monix-tail/shared/src/main/scala/monix/tail/internal/IterantFoldWhileLeft.scala]: I see, I've fixed it and added tests for this case - please see if I understand desired behavior correctly. :)

**Pull request merged**